File: 03-march/07-minimum-number-of-flips-to-make-the-binary-string-alternating/solution.java
class Solution {
    public int minFlips(String s) {
        int n = s.length();
        int diff0 = 0;
        int diff1 = 0;
        int ans = n;

        for (int i = 0; i < 2 * n; i++) {
            char ch = s.charAt(i % n);
            char expected0 = (i % 2 == 0) ? '0' : '1';
            char expected1 = (i % 2 == 0) ? '1' : '0';

            if (ch != expected0) {
                diff0++;
            }
            if (ch != expected1) {
                diff1++;
            }

            if (i >= n) {
                char left = s.charAt((i - n) % n);
                char leftExpected0 = ((i - n) % 2 == 0) ? '0' : '1';
                char leftExpected1 = ((i - n) % 2 == 0) ? '1' : '0';
                if (left != leftExpected0) {
                    diff0--;
                }
                if (left != leftExpected1) {
                    diff1--;
                }
            }

            if (i >= n - 1) {
                ans = Math.min(ans, Math.min(diff0, diff1));
            }
        }

        return ans;
    }
}

File: 03-march/07-minimum-number-of-flips-to-make-the-binary-string-alternating/solution.py
class Solution:
    def minFlips(self, s: str) -> int:
        n = len(s)
        diff0 = 0
        diff1 = 0
        ans = n

        for i in range(2 * n):
            ch = s[i % n]
            expected0 = "0" if i % 2 == 0 else "1"
            expected1 = "1" if i % 2 == 0 else "0"

            if ch != expected0:
                diff0 += 1
            if ch != expected1:
                diff1 += 1

            if i >= n:
                left = s[(i - n) % n]
                left_expected0 = "0" if (i - n) % 2 == 0 else "1"
                left_expected1 = "1" if (i - n) % 2 == 0 else "0"
                if left != left_expected0:
                    diff0 -= 1
                if left != left_expected1:
                    diff1 -= 1

            if i >= n - 1:
                ans = min(ans, diff0, diff1)

        return ans
